morpho5/geometry/functional.c

```diff
@@ -1343,6 +1343,133 @@ MORPHO_METHOD(FUNCTIONAL_TOTAL_METHOD, LinearElasticity_total, BUILTIN_FLAGSEMPT
 MORPHO_METHOD(FUNCTIONAL_GRADIENT_METHOD, LinearElasticity_gradient, BUILTIN_FLAGSEMPTY)
 MORPHO_ENDCLASS
 
+/* ----------------------------------------------
+ * Flory-Huggins
+ * ---------------------------------------------- */
+
+static value floryhuggins_aproperty;
+static value floryhuggins_bproperty;
+static value floryhuggins_cproperty;
+static value floryhuggins_phi0property;
+
+typedef struct {
+    objectmesh *refmesh;
+    grade grade;
+    double a,b,c;   // Flory-Huggins coefficients
+    value phi0; // Can be a number or a field
+} floryhugginsref;
+
+/** Prepares the reference structure from the object's properties */
+bool floryhuggins_prepareref(objectinstance *self, objectmesh *mesh, grade g, objectselection *sel, floryhugginsref *ref) {
+    bool success=false;
+    value refmesh=MORPHO_NIL, grade=MORPHO_NIL, phi0=MORPHO_NIL;
+    value a=MORPHO_NIL, b=MORPHO_NIL, c=MORPHO_NIL;
+
+    if (objectinstance_getproperty(self, linearelasticity_referenceproperty, &refmesh) &&
```

Comment:
This is quite a monster 'if' statement. For readability sake it might be worth making this a bool function.